### PR TITLE
Subscription Commands and Queries

### PR DESCRIPTION
**Pull request by timolegros**
## Link to Issue
Closes: #6946

## Description of Changes
- Creates trpc routes for all new subscription models
- Modifies core command logic to make routes secure by default
- Updated the `Comment.thread_id` attributes type from string to number to match the model definition + DB

## Test Plan
- Set `NEW_SUBSCRIPTION_API_FLAG` env var to true and play around with the routes from the swagger doc UI
- Interact with comments (create, update delete)

## Deployment Plan

1. 

## Other Considerations

-

### CODE REVIEW

**Comment by timolegros** [comment]
The PR is ready but #6916 must be merged first. Once it is merged I will take the PR out of draft.

File: libs/adapters/src/trpc/index.ts
Comment (Rotorsoft) [inline libs/adapters/src/trpc/index.ts:98]: shorter 😆 ?  `if(md.secure !== false || md.auth?.lengh)`

File: libs/core/src/schemas/commands/subscription.schemas.ts
Comment (Rotorsoft) [inline libs/core/src/schemas/commands/subscription.schemas.ts:18]: do you need a ticket to allow ZodEffect in commands?

File: libs/core/src/schemas/entities.schemas.ts
Comment (Rotorsoft) [inline libs/core/src/schemas/entities.schemas.ts:515]: PG_INT?

File: libs/model/src/community/index.ts
Comment (Rotorsoft) [inline libs/model/src/community/index.ts:1]: should we export samples?

File: libs/model/src/models/comment_subscriptions.ts
Comment (Rotorsoft) [inline libs/model/src/models/comment_subscriptions.ts:24]: are you using snake case in subscription schemas for some reason? ... most models use camel case with sequelize underscoring names to make it PG friendly... maybe we need to discuss a convention?

File: libs/model/src/subscription/GetSubscriptionPreferences.query.ts
Comment (Rotorsoft) [inline libs/model/src/subscription/GetSubscriptionPreferences.query.ts:15]: the proposed convention is to return an empty object or empty array or undefined instead of errors, depending on the expected results... the client app should decide what to do with empty results

File: libs/model/src/subscription/UpdateSubscriptionPreferences.command.ts
Comment (Rotorsoft) [inline libs/model/src/subscription/UpdateSubscriptionPreferences.command.ts:11]: guess upsert doesn't work here?

File: packages/commonwealth/server/api/index.ts
Comment (Rotorsoft) [inline packages/commonwealth/server/api/index.ts:14]: should we call this procedures or features?

File: libs/core/src/schemas/entities.schemas.ts
Comment (timolegros) [reply libs/core/src/schemas/entities.schemas.ts:515]: - `id` is generated by the DB
- `user_id` is a foreign key so it always has to be defined
- `thread_id` is also a foreign key

None of these values will or should be mocked in the DB so min/max is unnecessary. That said I will add `int()`

Comment (Rotorsoft) [reply libs/core/src/schemas/entities.schemas.ts:515]: @rbennettcw can give you some extra context, but he started adding min/max range to int cols to protect the db from out of range fields in payloads ... encapsulating this rule in something like PG_INT shouldn't hurt

File: libs/core/src/schemas/commands/subscription.schemas.ts
Comment (timolegros) [reply libs/core/src/schemas/commands/subscription.schemas.ts:18]: https://github.com/hicommonwealth/commonwealth/issues/7479

File: libs/model/src/community/index.ts
Comment (timolegros) [reply libs/model/src/community/index.ts:1]: No

File: libs/model/src/models/comment_subscriptions.ts
Comment (timolegros) [reply libs/model/src/models/comment_subscriptions.ts:24]: Not sure what you are referencing here. All of our model definitions use snake case. All of our Sequelize attribute types and entity schemas use snake case as well. Why would I cause camel case?

File: libs/model/src/subscription/GetSubscriptionPreferences.query.ts
Comment (timolegros) [reply libs/model/src/subscription/GetSubscriptionPreferences.query.ts:15]: Yes but again this ignores the fact that the data should exist when queried... I believe what we discussed was reporting the error to Rollbar but returning an empty result to the client? Otherwise, we silently fail and never know there is a failure since we don't report client errors to Rollbar.

File: libs/model/src/subscription/UpdateSubscriptionPreferences.command.ts
Comment (timolegros) [reply libs/model/src/subscription/UpdateSubscriptionPreferences.command.ts:11]: Technically unnecessary since there is a trigger which creates the SubscriptionPreference when a User is created and if a user updates their preference that assumes it already exists and can be viewed on the client.

File: packages/commonwealth/server/api/index.ts
Comment (timolegros) [reply packages/commonwealth/server/api/index.ts:14]: Aren't these just groups of trpc procedures?

File: libs/core/src/schemas/entities.schemas.ts
Comment (timolegros) [reply libs/core/src/schemas/entities.schemas.ts]: Makes sense.

File: libs/model/src/models/comment_subscriptions.ts
Comment (Rotorsoft) [reply libs/model/src/models/comment_subscriptions.ts:24]: 😱 true... snake case in all schemas it is... need to make sure I'm not using camel case in the new projection.

File: libs/model/src/subscription/GetSubscriptionPreferences.query.ts
Comment (Rotorsoft) [reply libs/model/src/subscription/GetSubscriptionPreferences.query.ts:15]: should we consider the option where users can try to query data that might not exist yet?

Comment (timolegros) [reply libs/model/src/subscription/GetSubscriptionPreferences.query.ts:15]: In the case when the data _should_ exist yes. If a user exists then their preferences must exist as well. Any deviation from that fact is an error/corruption.

File: libs/model/src/subscription/UpdateSubscriptionPreferences.command.ts
Comment (timolegros) [reply libs/model/src/subscription/UpdateSubscriptionPreferences.command.ts:11]: The problem is we currently don't strictly adhere to DDD and so Users can be created outside the Sequelize flow -> for example in the Discourse imports we don't use the Sequelize models which would mean new users would not have preferences when imported via Discourse.

I 100% agree it would be better to use Sequelize hooks for all triggers but it isn't an option for some of our models which have custom queries in which the hooks are not respected.

File: packages/commonwealth/server/api/index.ts
Comment (Rotorsoft) [reply packages/commonwealth/server/api/index.ts:14]: these are technically how we "group" endpoints to model artifacts... can be called artifacts ... procedures doesn't sound right

File: libs/model/src/subscription/GetSubscriptionPreferences.query.ts
Comment (Rotorsoft) [reply libs/model/src/subscription/GetSubscriptionPreferences.query.ts:15]: ok, so maybe we should create a utility guard to log the error and return empty... similar to mustExists but just to log?

Comment (timolegros) [reply libs/model/src/subscription/GetSubscriptionPreferences.query.ts:15]: I added a `shouldExist` guard which logs an InvalidState error (but doesn't throw) if the preferences don't exist

**Review by Rotorsoft — APPROVED** [review]

**Review by rbennettcw — APPROVED** [review]

**Pull request merged**
